stan/math/prim/mat/functor/mpi_parallel_call.hpp

```diff
@@ -0,0 +1,488 @@
+
+    const int num_local_jobs = local_job_params_dbl_.cols();
+    int local_outputs_per_job = num_local_jobs == 0 ? 0 : num_outputs_per_job_;
+    matrix_d local_output(
+        local_outputs_per_job == -1 ? 0 : local_outputs_per_job,
+        num_local_jobs);
+    std::vector<int> local_f_out(num_local_jobs, -1);
+
+    typename cache_x_r::cache_t& local_x_r = cache_x_r::data();
+    typename cache_x_i::cache_t& local_x_i = cache_x_i::data();
+
+    // check if we know already output sizes
+    if (cache_f_out::is_valid()) {
+      typename cache_f_out::cache_t& f_out = cache_f_out::data();
+      int num_outputs = 0;
+      for (int j = start_job; j < start_job + num_local_jobs; ++j)
+        num_outputs += f_out[j];
+      local_output.resize(Eigen::NoChange, num_outputs);
+    }
+
+    int local_ok = 1;
+    try {
+      for (int i = 0, offset = 0; i < num_local_jobs;
+           offset += local_f_out[i], ++i) {
+        const matrix_d job_output
+            = ReduceF()(local_shared_params_dbl_, local_job_params_dbl_.col(i),
+                        local_x_r[i], local_x_i[i], 0);
+        local_f_out[i] = job_output.cols();
+
+        if (local_outputs_per_job == -1) {
+          local_outputs_per_job = job_output.rows();
+          local_output.conservativeResize(local_outputs_per_job,
+                                          Eigen::NoChange);
+        }
+
+        if (local_output.cols() < offset + local_f_out[i])
+          local_output.conservativeResize(Eigen::NoChange,
+                                          2 * (offset + local_f_out[i]));
+
+        local_output.block(0, offset, local_output.rows(), local_f_out[i])
+            = job_output;
+      }
+    } catch (const std::exception& e) {
+      local_ok = 0;
+    }
+
+    // during first execution we distribute the output sizes from
+    // local jobs to the root. This needs to be done for the number of
+    // outputs of each result and the number of outputs per job.
+    if (num_outputs_per_job_ == -1) {
+      // before we can cache the sizes locally we must ensure
+      // that no exception has been fired from any node. Hence,
+      // share the info about the current status across all nodes.
+      int cluster_status = 0;
+      boost::mpi::reduce(world_, local_ok, cluster_status, std::plus<int>(), 0);
+      bool all_ok = cluster_status == static_cast<int>(world_size_);
+      boost::mpi::broadcast(world_, all_ok, 0);
+      if (!all_ok) {
+        // err out on the root
+        if (rank_ == 0) {
+          throw std::domain_error("MPI error on first evaluation.");
+        } else {
+          // and ensure on the workers that they return into their
+          // listening state
+          return result_t();
+        }
+      }
+
+      // execution was ok everywhere such that we can now distribute the
+      // outputs per job as recorded on the root on the first execution
+      boost::mpi::broadcast(world_, local_outputs_per_job, 0);
```

Comment:
Yes, there is. The root broadcasts it's value to everyone else. The workers overwrite the variable with the value from root. I cannot use the `num_outputs_per_job_` directly because it must be maintained at `-1` as long as all jobs from all workers were OK.